src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -348,6 +386,59 @@ private void EventLoop()
             }
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void ScheduleToProcessEvents()
+        {
+            // Schedule a thread pool work item to process events. Only one work item is scheduled at any given time to avoid
+            // over-parallelization. When the work item begins running, this field is reset to 0, allowing for another work item
+            // to be scheduled for parallelizing processing of events.
+            if (Interlocked.CompareExchange(ref _eventQueueProcessingRequested, 1, 0) == 0)
+            {
+                ThreadPool.UnsafeQueueUserWorkItem(this, preferLocal: false);
+            }
+        }
+
+        void IThreadPoolWorkItem.Execute()
+        {
+            // Indicate that a work item is no longer scheduled to process events. The change needs to be visible to enqueuer
+            // threads (only for EventLoop() currently) before an event is attempted to be dequeued. In particular, if an
+            // enqueuer queues an event and does not schedule a work item because it is already scheduled, and this thread is
+            // the last thread processing events, it must see the event queued by the enqueuer.
+            Interlocked.Exchange(ref _eventQueueProcessingRequested, 0);
+
+            ConcurrentQueue<SocketIOEvent> eventQueue = _eventQueue;
+            int startTimeMs = Environment.TickCount;
+            if (!eventQueue.TryDequeue(out SocketIOEvent ev))
+            {
+                return;
+            }
+
+            // An event was successfully dequeued, and there may be more events to process. Schedule a work item to parallelize
+            // processing of events, before processing more events. Following this, it is the responsibility of the new work
+            // item and the epoll thread to schedule more work items as necessary. The parallelization may be necessary here if
+            // the user callback as part of handling the event blocks for some reason that may have a dependency on other queued
+            // socket events.
+            ScheduleToProcessEvents();
+
+            while (true)
+            {
+                ev.Context.HandleEvents(ev.Events);
+
+                if (Environment.TickCount - startTimeMs >= 15)
```

Comment:
TickCount is sufficiently fast to call per work item?

Also, what is "15" here?  Is that just a reasonable quantum?  A comment would be helpful.

Reply:
It is probably a measurable perf difference with many measurements to weed out error, of maybe 2-3 K RPS, but that was more than countered by the change to remove the speculative work item scheduling, which can also has the effect of decreasing the rate of parallelization and its counteracting improvement is similarly difficult to measure. Roughly break-even between the two.

Reply:
I tried a bunch of values including using `Stopwatch` instead, considering that it is the max latency if such work items were to use up the whole thread pool budget, I figured it's a reasonable value. The average latency to run another queued work item as a result of this work item being long-running would be much less, something like `15 / proc count` in ideal circumstances, and even then it's kind of a corner case.

Reply:
I thought the comment I had following the break was sufficient, is there something specific that you would like to be mentioned in the comment?

Reply:
> is there something specific that you would like to be mentioned in the comment?

Just why you chose 15 specifically.  It's not clear to me from the comment.

Reply:
Oh ok, done

Reply:
You may need some sign coercion as TickCount overflows every 49'ish days of host uptime (so it will become -2.1billionish - +2.1billionish). Or ignore it if things remain correct-but-non-optimal for that once every 49 days event? :)

Reply:
That would only be possible if a user callback takes a vastly inordinate amount of time to run. `TickCount` overflow would not impact the check done here, the check done here is on the difference between current tick count and start tick count, that signed value would take a lot more to overflow, enough so that IMO it would already be incredibly unreasonable for user code to cause such a thing to happen for it to be worth considering here.

Reply:
But even if a single continuation did take more than 49 days, this check wouldn't hurt anything. This is just an optimization to try to avoid tying up a single thread pool thread by running too many small work items; if a user has a 49-day work item, this doesn't matter anyway.